splash/browser_tab.py

```diff
@@ -1116,6 +1193,176 @@ def log(self, message, min_level=None):
+
+
+class EventsStorage(QObject):
+    """
+    Object that allows to store JavaScript Event objects and access them.
+
+    This creates a JavaScript-compatible object (can be added to `window`)
+    that has `get_id()` function which can be called from JavaScript for
+    retrieving a unique id for each event object.
+
+    After adding to the JS window object the `init_storage(self)` method
+    should be called to initialize the storage. During the initialization
+    the storage object is connected to the QT signals which allows to call
+    appropriate methods of the specified event.
+    """
+    preventDefault = pyqtSignal(str)
+    stopImmediatePropagation = pyqtSignal(str)
+    stopPropagation = pyqtSignal(str)
+
+    def __init__(self, parent):
+        self.name = get_id()
+        super(EventsStorage, self).__init__(parent)
+
+    def init_storage(self):
+        frame = self.parent().web_page.mainFrame()
+        eval_expr = u"eval({})".format(escape_js("""
+        window[{storage_name}].events = {{}};
+
+        window[{storage_name}].callMethod = function(methodName) {{
+            return function(eventId) {{
+                var storage = window[{storage_name}].events;
+                storage[eventId][methodName].call(storage[eventId]);
+            }};
+        }}
+
+        window[{storage_name}].preventDefault.connect(window[{storage_name}].callMethod('preventDefault'))
+        window[{storage_name}].stopImmediatePropagation.connect(window[{storage_name}].callMethod('stopImmediatePropagation'))
+        window[{storage_name}].stopPropagation.connect(window[{storage_name}].callMethod('stopPropagation'))
+
+        window[{storage_name}].add = function(event) {{
+            var id = window[{storage_name}].getId()
+            window[{storage_name}].events[id] = event;
+            return id;
+        }}
+        """.format(storage_name=escape_js(self.name))))
+
+        frame.evaluateJavaScript(eval_expr)
+
+    @pyqtSlot(name="getId", result=str)
+    def get_id(self):
+        return get_id()
+
+    def get_event_property(self, event_id, property_name):
+        frame = self.parent().web_page.mainFrame()
+        eval_expr = u"eval({})".format(escape_js("""
+        window[{storage_name}].events[{event_id}][{property_name}]
+        """.format(
+            storage_name=escape_js(self.name),
+            event_id=escape_js(event_id),
+            property_name=escape_js(property_name)
+        )))
+
+        result = frame.evaluateJavaScript(get_process_errors_js(eval_expr))
+
+        return result.get('result', None)
+
+    def remove_event(self, event_id):
+        frame = self.parent().web_page.mainFrame()
+        eval_expr = u"eval({})".format(escape_js("""
+        delete window[{storage_name}].events[{event_id}]
+        """.format(
+            storage_name=escape_js(self.name),
+            event_id=escape_js(event_id),
+        )))
+
+        frame.evaluateJavaScript(get_process_errors_js(eval_expr))
```

Comment:
it looks like `eval_expr = "eval({})".format(escape_js(txt)); result = frame.evaluateJavaScript(get_process_errors_js(eval_expr)` is repeated a few time; what do you think about introducing a helper function for that? Or just for `get_process_errors_js(eval("{}".format(escape_js(txt))))`?